packages/jest-mock/src/index.ts

```diff
@@ -135,6 +141,12 @@ export interface MockInstance<T extends FunctionLike = UnknownFunction> {
   mockRestore(): void;
   mockImplementation(fn: T): this;
   mockImplementationOnce(fn: T): this;
+  withImplementation<R extends WithImplementationCallbackReturn>(
+    fn: T,
+    callback: () => R,
+  ): R extends WithImplementationAsyncCallbackReturn
+    ? Promise<void>
+    : undefined;
```

Comment:
Just an idea. What about function overloads? Isn’t it better pattern in this case? I was trying to play with them briefly and seems like this could help getting rid of `@ts-expect-error`. Or you tried and bump into some issue?

```suggestion
  withImplementation(fn: T, callback: () => void): void;
  withImplementation(fn: T, callback: () => Promise<unknown>): Promise<void>;
```

Reply:
I did not come across this solution when working on the problem, I'll try it out and see what happens :+1:

Reply:
It appears to be working as intended :+1: